### PR TITLE
Remove Arc<RwLock<>> from Wallet

### PR DESCRIPTION
**Pull request by delta1**
## Description

Removes the unnecessary `Arc<RwLock<_>>` from the Wallet, and updates the cascading changes. The reasoning for removal is that the Wallet object should already be a thread-safe and lightweight collection of handles, and so be easily cloned when necessary. 

For the Wallet to derive `Clone`, it is required that the WalletBackend and ContactsBackend implementors are also Clone. Additionally, the `Shutdown` type previously held in the Wallet is hoisted up a level into the client applications creating the wallet.  

## Motivation and Context

From feedback in [#2291](https://github.com/tari-project/tari/pull/2291#discussion_r503287026) 

## How Has This Been Tested?

- [x] cargo test
- [x] manually

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 1]
```diff
@@ -7,33 +6,24 @@ use tari_wallet::{
...
-                TransactionServiceSqliteDatabase,
-                OutputManagerSqliteDatabase,
-                ContactsServiceSqliteDatabase,
-            >,
-        >,
+    wallet: Wallet<
+        WalletSqliteDatabase,
+        TransactionServiceSqliteDatabase,
+        OutputManagerSqliteDatabase,
+        ContactsServiceSqliteDatabase,
```
Comment (philipr-za) [inline applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs]:
```suggestion
    wallet: WalletSqlite
```

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 2]
```diff
@@ -7,33 +6,24 @@ use tari_wallet::{
...
-                    TransactionServiceSqliteDatabase,
-                    OutputManagerSqliteDatabase,
-                    ContactsServiceSqliteDatabase,
-                >,
-            >,
+        wallet: Wallet<
+            WalletSqliteDatabase,
+            TransactionServiceSqliteDatabase,
+            OutputManagerSqliteDatabase,
+            ContactsServiceSqliteDatabase,
```
Comment (philipr-za) [inline applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs]:
```suggestion
        wallet: WalletSqlite
```

File: applications/tari_console_wallet/src/main.rs [hunk 3]
```diff
@@ -43,17 +43,23 @@ mod utils;
 
 /// Application entry point
 fn main() {
-    match main_inner() {
-        Ok(_) => std::process::exit(0),
+    let mut shutdown = Shutdown::new();
+
+    match main_inner(shutdown.to_signal()) {
+        Ok(_) => {
+            shutdown.trigger().unwrap();
```
Comment (philipr-za) [inline applications/tari_console_wallet/src/main.rs]: After we trigger the shutdown we need to wait on the wallet to complete its shutdown (which really just waits for the comms stack to finish its shutdown). 

I recommend we change the wallets shutdown (copied below) to a wait_until_shutdown so we let the client do the triggering.

```Rust
pub async fn wait_until_shutdown(mut self) {
        self.comms.wait_until_shutdown().await;
    }
```

Also whenever you are using an unwrap it is better to use an .expect("...") with a message saying what was supposed to happen.

File: applications/tari_console_wallet/src/ui/app.rs [hunk 4]
```diff
@@ -60,15 +60,11 @@ pub const LOG_TARGET: &str = "wallet::ui::app";
...
-                TransactionServiceSqliteDatabase,
-                OutputManagerSqliteDatabase,
-                ContactsServiceSqliteDatabase,
-            >,
-        >,
+    pub wallet: Wallet<
+        WalletSqliteDatabase,
+        TransactionServiceSqliteDatabase,
+        OutputManagerSqliteDatabase,
+        ContactsServiceSqliteDatabase,
```
Comment (philipr-za) [inline applications/tari_console_wallet/src/ui/app.rs]:
```suggestion
    pub wallet: WalletSqlite
```

File: applications/tari_console_wallet/src/ui/app.rs [hunk 5]
```diff
@@ -80,15 +76,11 @@ impl<B: Backend> App<B> {
...
-                    TransactionServiceSqliteDatabase,
-                    OutputManagerSqliteDatabase,
-                    ContactsServiceSqliteDatabase,
-                >,
-            >,
+        wallet: Wallet<
+            WalletSqliteDatabase,
+            TransactionServiceSqliteDatabase,
+            OutputManagerSqliteDatabase,
+            ContactsServiceSqliteDatabase,
```
Comment (philipr-za) [inline applications/tari_console_wallet/src/ui/app.rs]:
```suggestion
        wallet: WalletSqlite
```

File: base_layer/wallet/src/wallet.rs [hunk 6]
```diff
@@ -221,8 +220,8 @@ where
 
     /// This method consumes the wallet so that the handles are dropped which will result in the services async loops
     /// exiting.
-    pub async fn shutdown(mut self) {
-        self.shutdown.trigger().expect(
+    pub async fn shutdown(self, mut shutdown: Shutdown) {
```
Comment (philipr-za) [inline base_layer/wallet/src/wallet.rs]: I propose that we rather change this into a wait_for_shutdown(...) and have the client's trigger the shutdown. So in the FFI lib you will need to do the following (In the Wallet FFI case we do not want to panic if the shutdown triggers are wrong so not using an expect() there)

```Rust
pub unsafe extern "C" fn wallet_destroy(wallet: *mut TariWallet) {
    if !wallet.is_null() {
        let mut m = Box::from_raw(wallet);
        match m.shutdown.trigger() {
            Err(_) => error!(target: LOG_TARGET, "No one is listening for the shutdown trigger!"),
            Ok(()) =>  m.runtime.block_on(m.wallet.wait_for_shutdown()),
            
        }
    }
}
```
And the equivalent in the console wallet and also the base node while there is still a wallet in there.

**Review by philipr-za — COMMENTED** [review]
Getting there, you will also need to wire the Base Node shutdown into its wallet.

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs (hunk 1 above)
Comment (delta1) [reply applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs]: Thanks was just busy with this

File: applications/tari_console_wallet/src/main.rs (hunk 3 above)
Comment (delta1) [reply applications/tari_console_wallet/src/main.rs]: Excellent thanks will do

File: base_layer/wallet/src/wallet.rs (hunk 6 above)
Comment (delta1) [reply base_layer/wallet/src/wallet.rs]: Will do

**Comment by delta1** [comment]
> Getting there, you will also need to wire the Base Node shutdown into its wallet.

~Cool will do - will have to investigate how the BN is starting its wallet~ 

As discussed, Base Node uses the WalletBootstrap and already wires up a shutdown signal.

**Review by stringhandler — APPROVED** [review]

**Comment by delta1** [comment]
Thanks for the merge but I think they may be an issue with the tests...

**Pull request merged**